File: F#/chap07/Chap0707.fs
namespace Chap0707

open Chap0705.C070502
open Chap0706

type AsyncResult<'success, 'failure> = Async<Result<'success, 'failure>>

type CheckProductCodeExists = Undefined
type CheckAddressExists = Undefined
type UnvalidatedOrder = Undefined
type ValidatedOrder = Undefined
type ValidationError = Undefined
type GetProductPrice = Undefined
type PricedOrder = Undefined
type PricingError = Undefined

module PatternA =
    type ValidateOrder =
        CheckProductCodeExists // 依存を明示
            -> CheckAddressExists // 依存を明示
            -> UnvalidatedOrder // In
            -> AsyncResult<ValidatedOrder, ValidationError list> // Out

    type PriceOrder =
        GetProductPrice // 依存を明示
            -> ValidatedOrder // In
            -> Result<PricedOrder, PricingError> // Out

module PatternB =
    type ValidateOrder =
        UnvalidatedOrder // In
            -> AsyncResult<ValidatedOrder, ValidationError list> // Out

    type PriceOrder =
        ValidatedOrder // In
            -> Result<PricedOrder, PricingError> // Out


type PlaceOrder = Undefined
type PlaceOrderEvent = Undefined
type PlaceOrderError = Undefined

type PlaceOrderWorkflow =
    PlaceOrder // In
        -> AsyncResult<PlaceOrderEvent list, PlaceOrderError> // Out

File: Scala/chap07/src/main/scala/Chap0707.scala
import cats.data.EitherT

import scala.concurrent.Future

object Chap0707:

  case class CheckProductCodeExists(value: Any)

  case class CheckAddressExists(value: Any)

  case class UnvalidatedOrder(value: Any)

  case class ValidatedOrder(value: Any)

  case class ValidationError(value: Any)

  case class GetProductPrice(value: Any)

  case class PricedOrder(value: Any)

  case class PricingError(value: Any)

  object PatternA:
    type ValidateOrder =
      CheckProductCodeExists // 依存
        => CheckAddressExists // 依存
        => UnvalidatedOrder // In
        => EitherT[Future, ValidationError, ValidatedOrder] // Out

    type PriceOrder =
      GetProductPrice // 依存
        => ValidatedOrder // In
        => Either[PricingError, PricedOrder] // Out

  object PatternB:
    type ValidateOrder =
      UnvalidatedOrder // In
        => EitherT[Future, ValidationError, ValidatedOrder] // Out

    type PriceOrder =
      ValidatedOrder // In
        => Either[PricingError, PricedOrder] // Out

  case class PlaceOrder(value: Any)

  case class PlaceOrderEvent(value: Any)

  case class PlaceOrderError(value: Any)

  type PlaceOrderWorkflow =
    PlaceOrder // In
      => EitherT[Future, PlaceOrderError, List[PlaceOrderEvent]]

end Chap0707
